drivers/gpio/gpio_mcp23s17.c

```diff
@@ -0,0 +1,478 @@
+/*
+ * Copyright (c) 2020 Geanix ApS
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file Driver for MCP23S17 SPI-based GPIO driver.
+ */
+
+#include <errno.h>
+
+#include <kernel.h>
+#include <device.h>
+#include <init.h>
+#include <sys/byteorder.h>
+#include <drivers/gpio.h>
+#include <drivers/spi.h>
+
+#include "gpio_utils.h"
+#include "gpio_mcp23s17.h"
+
+#define LOG_LEVEL CONFIG_GPIO_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(gpio_mcp23s17);
+
+/**
+ * @brief Read both port 0 and port 1 registers of certain register function.
+ *
+ * Given the register in reg, read the pair of port 0 and port 1.
+ *
+ * @param dev Device struct of the MCP23S17.
+ * @param reg Register to read (the PORTA of the pair of registers).
+ * @param buf Buffer to read data into.
+ *
+ * @return 0 if successful, failed otherwise.
+ */
+static int read_port_regs(struct device *dev, u8_t reg, u16_t *buf)
+{
+	struct mcp23s17_drv_data *const drv_data =
+		(struct mcp23s17_drv_data *const)dev->driver_data;
+	int ret;
+	u16_t port_data;
+
+	u8_t addr = MCP23S17_ADDR | MCP23S17_READBIT;
+	u8_t buffer_tx[4] = { addr, reg, 0, 0 };
+
+	const struct spi_buf tx_buf = {
+		.buf = buffer_tx,
+		.len = 4,
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1,
+	};
+	const struct spi_buf rx_buf[2] = {
+		{
+			.buf = NULL,
+			.len = 2
```

Comment:
I guess this should be `len = 0` since buf is NULL?

Reply:
Hmm, I could be wrong, but this is the 2 bytes that will be shifted in when the read command is shifted out.
Also seen here:
https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/sensor/lsm6dsl/lsm6dsl_spi.c#L50

Reply:
Ok, got it. I (naively) assumed the `len` was only used for an indication for how many bytes are behind the `buf` pointer.